File: java_solutions/142.linked-list-cycle-ii.java
/*
 * @lc app=leetcode id=142 lang=java
 *
 * [142] Linked List Cycle II
 */

public class Solution {
    public ListNode detectCycle(ListNode head) {
        return java.util.stream.Stream.of(new java.util.IdentityHashMap<ListNode, Boolean>()).flatMap(v -> java.util.stream.Stream.<ListNode>iterate(head, n -> n != null, n -> n.next).filter(n -> v.put(n, true) != null)).findFirst().orElse(null);
    }
}

File: kotlin_solutions/142.linked-list-cycle-ii.kt
/*
 * @lc app=leetcode id=142 lang=kotlin
 *
 * [142] Linked List Cycle II
 */

class Solution {fun detectCycle(head: ListNode?): ListNode? = hashSetOf<ListNode>().run {generateSequence(head) { it?.next }.firstOrNull { !add(it) }}}
